Implement delete/max comments feature

vedantroy opened the pull request:
Finished the final part of the week 3 tutorial, [here](https://github.com/vedantroy/step-2020/blob/master/walkthroughs/week-3-server/step-6-additional-features-walkthrough.md). @sooniln @alexanderkuznetsov96

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -54,13 +56,21 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
-    ArrayList<String> comments = new ArrayList<>();
-    for (Entity entity : results.asIterable()) {
-      comments.add((String)entity.getProperty("value"));
+    String maxCommentsString = request.getParameter("max_comments");
```
What might be more reasonable behavior in the event the query does not supply a max_comments key, or supplies a malformed one?

vedantroy replied:
Will change the servlet to return a error with return code 400.

sooniln replied:
Returning an error is not un-reasonable, so I would be fine with returning an error. However, what I'd like to consider more broadly is the distinction between recoverable and non-recoverable errors, as well as thinking in terms of how client's might like to use an API.

Let's say you're a client making a request to a data API like this - what behavior would you expect if you don't include the max_comments value? Consider the similar API PreparedQuery exposes, what choices have devs made for when clients elide a similar parameter? As a general rule of thumb, you want APIs to be strict when enforcing correctness, and loose in terms useability (as useable as possible, awkward phrasing aside). Does not passing in the max_comments value affect the ability for your API to be correct, and does it expose any surprising behavior to the user?

Yet another angle on the question: What's best for the user? From my perspective an average user clearly wouldn't care what the exact value of max_comments is, and asking them to care seems unreasonable in the majority of cases. So then someone else has to care - and that leaves only the client or the server. The question is then, who is better placed to make the decision? When making a decision, the client has user specific information, and client information, but not server information. The server has user specific information, client specific information, and server information, so the server has the most information if a decision needs to be made. Arguably, the server is thus the most natural place to make a decision on what should happen if the user doesn't care about the value of max_comments.

If you think through these cases, and believe that the responsibility should be with the client (ie, forcing the client to include max_comments), then I'm fine with that - I want to see your reasoning however :)

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -35,17 +36,18 @@
 public class DataServlet extends HttpServlet {
 
   private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private final String COMMENT_TAG = "comment";
```
As this appears to be a constant, should it be static? If so, java style would say to put constants at the top of your file generally.

Re the naming, is this a tag? Or better considered a key?

Re consistency, this string literal is extracted into a constant, but max_comments below isn't?

vedantroy replied:
Re: naming. I guess this should be considered a "kind". That is the official naming in the Firebase documentation. (Entities can have "kinds").

Re: consistency. I extracted this string literal because it was used in multiple places, but `max_comments` was not. Should I also extract `max_comments`?

sooniln replied:
"kind" sounds good to me. The general test for "should X be a constant" should rarely be "in how many places is this used". When this is compiled, the code/memory code will be exactly the same whether you use the same string literal everywhere, or extract into a constant. I would ask, what is the most consistent, and thus the easiest for someone else reading my code to understand?

When reading through this, when I see one is extracted into a constant, and the other isn't, why first thought is, "this seems odd, am I missing some piece of information in reading this"? If I saw that you had used always a string literal, or always a constant, I would attribute it to stylistic choice, and not think too much of it. One of the highest priorities for all code should be to be as readable as possible - and that often requires thinking like someone who is completely new to a codebase, and may not have background on why you made choices you did.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -54,13 +56,21 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
+    String maxCommentsString = request.getParameter("max_comments");
+    if (maxCommentsString != null) {
+      Integer maxComments = Integer.parseInt(maxCommentsString);
+      Query query = new Query(COMMENT_TAG);
```
minor nit: you are tying your database key and query key to the same name by using the same constant. fine for now, but something to think about in terms of how code might change in the future.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -54,13 +56,21 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
+      for (Entity entity : results.asIterable()) {
+        if (i++ == maxComments) break;
+        comments.add((String)entity.getProperty("value"));
+      }
```
Is there a potentially more efficient method of retrieving N comments available through the PreparedQuery APIs?

vedantroy replied:
Addressed

vedantroy commented:
@sooniln. I addressed some of the comments, I left the server in its current state with respect to returning a 400 error code if the `max_comments` query parameter was not specified, however I can change that behavior.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -35,17 +37,18 @@
 public class DataServlet extends HttpServlet {
 
   private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final String COMMENT_KIND = "comment";
```
nit: in java style, static members usually come before non-static members.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -54,13 +57,20 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
+      Integer maxComments = Integer.parseInt(maxCommentsString);
+      Query query = new Query(COMMENT_KIND);
+      List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments));
+      String[] comments = results
```
Nice usage of the stream APIs. If I were using this in prod I might want to do a sanity check on any memory/CPU increase, but the reduced verbosity is very nice to have.

sooniln commented on portfolio/src/main/webapp/script.js:
```diff
@@ -45,12 +45,38 @@ content.appendChild(
-      p.textContent = comment;
-      document.body.append(p);
+
+let maxComments = 3;
```
I'm not as familiar with JavaScript best practices and constraints, but should maxComments be an argument to getComments() instead of a global?

vedantroy commented:
@sooniln. Addressed the changes. I considered what you said about what reasonable behavior and made the server return supply a default value if the `max_comments` query parameter was malformed/missing.

With regards to the extraction of string literals to constants, my previous intuition on this was "extract literals into `static final` variables if you are repeating the string literal/numeric literal in multiple places because you want to avoid repetition. Otherwise, leave it alone." But I guess it's better to take a stance and stick to it within the codebase.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -26,14 +27,22 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private static final String JSON_CONTENT_TYPE = "application/json;";
```
nit: when naming constants like this it's often helpful to order the various name components from least specific to most specific. ie:

CONTENT_TYPE_JSON instead of JSON_CONTENT_TYPE
KIND_COMMENT instead of COMMENT_KIND
etc...

the motivation is that it's easier for readers to see groupings and understand context. for example:

KIND_COMMENT = ...
KIND_ABC = ...
KIND_Z = ...

it's easy for the reader to see 1) that these are related in the same category 2) what the distinction between each is (since the text will naturally line up more nicely). vs:

COMMENT_KIND = ...
ABC_KIND = ...
Z_KIND = ...

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -42,25 +51,33 @@ public DataServlet() {
-    for (Entity entity : results.asIterable()) {
-      comments.add((String)entity.getProperty("value"));
+    String maxCommentsString = request.getParameter(MAX_COMMENTS_PARAM);
+    Integer maxComments = null;
```
maxComments can be an int instead of Integer? This is more natural, and can be achieved via:

int maxComments = DEFAULT_MAX_COMMENTS;
try {
  maxComments = Integer.parseInt(maxCommentsString);
} catch (NumberFormatException e) {
  // do nothing (maybe log an error in prod, who knows)
}

sooniln commented on portfolio/src/main/webapp/script.js:
```diff
@@ -45,12 +45,40 @@ content.appendChild(
+form.addEventListener("submit", (e) => {
+  e.preventDefault();
+  const data = new FormData(form);
+  for (const [key, val] of data) {
```
it looks like this loop will continue iterating (well, if there were hypothetically more elements in the form) even after work is done? perhaps a more natural way of doing what you want, if i'm understanding correctly, would be to extract the key you care about, and use that directly? ie:

const maxComments = data.get("max_comments") // or default?
getComments(maxComments)

vedantroy commented:
@sooniln Addressed the comments.

sooniln commented on portfolio/src/main/webapp/script.js:
```diff
@@ -45,12 +45,36 @@ content.appendChild(
+  const data = new FormData(form);
+  const newMaxComments = parseInt(data.get("max-comments"));
+  if (newMaxComments !== maxComments && newMaxComments !== NaN) {
+    maxComments = parseInt(newMaxComments);
```
newMaxComments is already derived from calling parseInt (line 54), is it necessary to call it again?
